onnxruntime/core/providers/cpu/rnn/rnn_helpers.h

```diff
@@ -230,17 +230,49 @@ void ExecuteLambdaInParallel(const std::string& name, TLambda lambda, int max, i
   ORT_UNUSED_PARAMETER(name);
   ORT_UNUSED_PARAMETER(logger);
 
-  std::atomic<int> done(0);
-  for (int i = 0; i < max; i += step) {
-    ttp.Schedule([lambda, i, &done]() {
-      lambda(i);
-      ++done;
+  // ORT_ENFORCE may and does throw at times from within the tasks that run
+  // on a thread-pool. Without propagating exceptions the process exits silently
+  // which will make diagnosing bugs more difficult.
+
+  // \! UGLY
+  // We have a problem here with the current thread-pool is that it takes std::function
+  // by value and copies it more than once (even though it is movable).
+  //
+  // To report status and exceptions properly it's better to use
+  // futures and promises but they are not copyable, so we can't come up with a functor
+  // with a promise member and we are downgrading to C++11 where we can't have captures that moved in.
+  //
+  // At the same time promises MUST live in the child thread so if we throw from the main thread
+  // we don't destroy any promises that are on the main thread stack which children threads may still be using.
+  //
+  // The only solution with the current Eigen that comes to mind is to have shared_ptr to with std::promise.
+  //
+
+  std::vector<std::future<void> > futures;
```

Comment:
Can this vector be reserve()'d?

Reply:
Sure.

---
In reply to: [327850443](https://github.com/microsoft/onnxruntime/pull/1905#discussion_r327850443) [](ancestors = 327850443)